File: solutions/070_easy/go/n_common_words.go
package main

import (
	"fmt"
    "log"
	"io/ioutil"
	"os"
	"sort"
	"strconv"
	"strings"
)

type WordCount struct {
	Word  string
	Count uint
}

// Define type to sort WordCounts
type ByCount []WordCount
func (a ByCount) Len() int           { return len(a) }
func (a ByCount) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a ByCount) Less(i, j int) bool { return a[i].Count < a[j].Count }


// Return a mapping of words (strings without whitespace) to their frequency in
// the string provided
func wordCount(s string) map[string]uint {
	m := make(map[string]uint)
	words := strings.Fields(s)
	for _, word := range words {
		if v, present := m[word]; !present {
			m[word] = v
		}
		m[word] += 1
	}
	return m
}

func main() {
    // Read file provided as first argument
	fname := os.Args[1]
	file_bytes, err := ioutil.ReadFile(fname)
	if err != nil {
            log.Fatal("Failed to load file: ", err)
	}
	file_text := string(file_bytes)

    // Read n and parse as uint
	n_str := os.Args[2]
	n, err := strconv.ParseUint(n_str, 10, 32)
	if err != nil {
            log.Fatal("Failed to parse n: ", err)
	}

    // Count words and sort these counts
    word_count := wordCount(file_text)
	counts := []WordCount{}
	for word, count := range word_count {
		counts = append(counts, WordCount{word, count})
	}
	sort.Sort(sort.Reverse(ByCount(counts)))

    last_ind := int(n)
    if last_ind > len(counts) {
            fmt.Println("NOTE: File has fewer than ", n, " unique words.")
            fmt.Println("      Only able to display ", len(counts), " words.")
            last_ind = len(counts)
    }
	for i := 0; i < last_ind; i++ {
		fmt.Println(counts[i])
	}
}

File: solutions/070_easy/rust/src/n_common_words.rs
use std::env;
use std::io;
use std::io::prelude::*;
use std::fs::File;
use std::collections::HashMap;

fn main() {
    // read in file name and n
    let mut argv = env::args();
    let usage = format!("USAGE: ./{} {{file_name}} {{top_n_words}}", argv.next().unwrap());
    let fname: String = argv.next().expect(&usage);
    let n_str: String = argv.next().expect(&usage);
    let parse_int_err = format!("Failed to parse arg '{}' to an int", n_str);
    let n: usize = n_str.parse::<usize>().ok().expect(&parse_int_err);

    // read in file
    let file_open_err = format!("Failed to open {}", fname);
    let file = File::open(&fname)
                    .ok()
                    .expect(&file_open_err);
    let freader = io::BufReader::new(&file);

    // break file into words
    let lines: Vec<String> = freader.lines()
        .map(|line_result| line_result.ok().unwrap())
        .collect();
    let words = lines.iter()
        .flat_map(|line| line.split(" "))
        .filter(|word| !word.is_empty())
        .map(|word| word.to_string());

    // count words
    let mut word_count: HashMap<String, i32> = HashMap::new();
    for i in words {
        let count = word_count.entry(i).or_insert(0);
        *count += 1;
    }

    // Sort mapping
    let mut pairs: Vec<(&String, &i32)> = word_count.iter().collect();
    pairs.sort_by(|a, b| b.1.cmp(a.1));

    // Output top n
    if pairs.len() < n {
            println!("NOTE: File has fewer than {} unique words.", n);
            println!("      Only able to display {} words.", pairs.len());
    }
    for (num, pair) in pairs.iter().enumerate() {
        if num >= n {
            break;
        }
        println!("{}", pair.0)
    }
}
